Source/Core/AudioCommon/OpenALStream.cpp

```diff
@@ -149,10 +155,13 @@ void OpenALStream::SoundLoop()
   memset(uiBuffers, 0, numBuffers * sizeof(ALuint));
   uiSource = 0;
 
-  // Checks if a X-Fi is being used. If it is, disable FLOAT32 support as this sound card has no
-  // support for it even though it reports it does.
+  if (alIsExtensionPresent("AL_EXT_float32"))
+    float32_capable = true;
+
+  // Checks if a X-Fi is being used. If it is, enable FIXED32 support as this sound card
+  // is the only one that supports it for now.
   if (strstr(alGetString(AL_RENDERER), "X-Fi"))
-    float32_capable = false;
+    fixed32_capable = true;
```

Comment:
Actually, the X-Fi driver never reported an error when buffering 32-bit floating point data (OpenAL was never able to fall back to 16-bit fixed point buffers), so it played this data as 32-bit fixed point creating disgusting noises. 
I was doing it wrong before, I should have checked for the extensions instead of forcing any X-Fi to not use 32-bit floating point buffers:
```
if (!alIsExtensionPresent("AL_EXT_float32"))
    float32_capable = false;
```

But now I prefer having the 32-bit floating point support normally off, and just checking if the OpenAL driver has this extension to enable support.
The X-Fi driver never supported 32-bit floating point data, but has undocumented support for 32-bit fixed point. There is no extension to check for support, but I believe all X-Fi with hardware OpenAL supports this feature.

"OpenAL Soft" supports `AL_EXT_float32` so it works fine, "Generic Software" doesn't so it falls back to 16-bit fixed point.